include/lbann/layers/io/input/generic_input_layer.hpp

```diff
@@ -848,21 +915,26 @@ class generic_input_layer : public io_layer {
   }
 
  protected:
+  std::vector<generic_io_buffer*> m_io_buffers;
+  std::atomic<int> m_active_buffer;
+
   dataset m_training_dataset;
   dataset m_testing_dataset;
   dataset m_validation_dataset;
   //  bool m_data_sets_span_models;
 
   data_reader_map_t m_data_readers;
  //  std::map<execution_mode, dataset_stats> m_dataset_stats;
+  bool m_data_set_processed;
+  std::mutex dr_mutex;
 };
 
 template<> inline void generic_input_layer::initialize_io_buffer<partitioned_io_buffer>(lbann_comm *comm, int num_parallel_readers, std::map<execution_mode, generic_data_reader *> data_readers) {
-  io_buffer = new partitioned_io_buffer(comm, num_parallel_readers, data_readers, m_expected_num_child_layers);
+  m_io_buffers.push_back(new partitioned_io_buffer(comm, num_parallel_readers, data_readers, m_expected_num_child_layers));
```

Comment:
Why are these specialized? They appear to be the same functions with a different `BufferT` being `new`-ed in each with the same arguments. Am I missing something?

Reply:
Because the I/O buffer is very different between the partitioned and distributed versions.  Also the distributed I/O buffer is going to be deprecated soon.

Reply:
That's fine, but I mean "why is the source code duplicated when all that differs is the type (but not the arguments)?" Should this and the next function not just be collapsed into the single function template:
```cpp
template <typename T> inline void generic_input_layer::initialize_io_buffer(
    lbann_comm *comm, int num_parallel_readers,
    std::map<execution_mode, generic_data_reader *> data_readers) {
  m_io_buffers.push_back(new T{comm, num_parallel_readers, data_readers, m_expected_num_child_layers});
}
```

Reply:
Also, do you mean to do a deep copy of the `std::map` object there? You're doing at least 2, one when you enter `initialize_io_buffer` and one when in the ctor of your `{partitioned|distributed}_io_buffer`. If you need a clean copy to go into the buffer, the function should *actually* be
```cpp
template <typename T> inline void generic_input_layer::initialize_io_buffer(
    lbann_comm *comm, int num_parallel_readers,
    std::map<execution_mode, generic_data_reader *> data_readers) {
  m_io_buffers.push_back(new T{comm, num_parallel_readers, std::move(data_readers), m_expected_num_child_layers});
}
```
so that you only do the one copy that you need and then move the clean copy into the buffer. If, on the other hand, you don't know if the buffer needs a clean copy, the function should be written:
```cpp
template <typename T> inline void generic_input_layer::initialize_io_buffer(
    lbann_comm *comm, int num_parallel_readers,
    std::map<execution_mode, generic_data_reader *> const& data_readers) {
  m_io_buffers.push_back(new T{comm, num_parallel_readers, data_readers, m_expected_num_child_layers});
}
```
so that the buffer constructor can decide for itself if it needs a clean copy and it is then responsible for constructing it.

Reply:
I think that we can fix this up after some code cleanup in a future PR.